woocommerce/woocommerce-ios
# Login: Show error screens upon app start if there was error during site login address flow previously

Closes: #7419


### Description

Previously, when a user is logged in without a selected site, we're showing the store picker. In many cases, they end up with an empty store picker with an action button to install Jetpack. This screen can be misleading and is not helpful.

This PR improves this experience by adding a few checks:
- If the user encountered an issue with the site address login flow earlier, we keep the site URL in the local storage. So the next time they start the app, we check for the problem of the app and display an error screen to help them pick up where they left off (e.g account mismatched or missing Woo).
- If the user doesn't have any connected store, we log them out since the empty store picker isn't helpful. We can consider showing some other content for store creation here for further enhancement later.

Technical changes:
- Update `LoggedOutAppSettings` with a new variable for `errorLoginSiteAddress`.
- Inject `LoggedOutAppSettings` to `AuthenticationManager` from `AppCoordinator`. Since `AuthenticationManager` is a singleton, and I don't want to move `LoggedOutAppSettings` to `ServiceLocator` as well, so I use method injection here instead of initializer injection.
- Extract a method for getting the error view controller for a given site URL in `AuthenticationManager`. This method is internal so that it can be accessed by `AppCoordinator`.
- Update `AppCoordinator` to present error view controller or log out if needed when handling the logged-in state without default store.

### Testing instructions

Testing Account Mismatch error:
1. Log out of the app if needed, and select Enter site address.
2. Enter the address for a Woo store to which you don't have ac
cess.
3. Proceed to log in with your WP.com account.
4. After the login succeeds, the Account Mismatch screen is presented. Kill the app.
5. Open the app again, notice that you get the Account Mismatch screen again. Tap Log out, you should get back to the login screen.

Testing No Woo error:
1. Log out of the app if needed, and select Enter site address.
2. Enter the address of a self-hosted site with Jetpack and no Woo.
3. Proceed to log in with your WP.com account.
4. After the login succeeds, the No Woo screen is presented. Kill the app.
5. Open the app again, notice that you get the No Woo screen again. The Install WooCommerce and Log out buttons should still work as expected.

The last case where user is logged in without neither of the above errors and has no connected store is probably an edge case. Please feel free to test it by the following steps:
- Invite a test user to your store.
- Follow the site address login flow to log in with that user.
- On the store picker screen, kill the app.
- Remove the test user from your store.
- Open the app again, the user should stay in the prologue screen now.

### Screenshots

Account Mismatch error:
<img src="https://user-images.githubusercontent.com/5533851/183052618-25c6ddab-4e66-4779-94bc-d89f7ee71ae6.png" width=320 />

No Woo error:
<img src="https://user-images.githubusercontent.com/5533851/183052700-f2063d99-7164-4428-90e1-cb0c509a1965.png" width=320 />



---
